docs/admin/ServerSideKeygen.adoc

```diff
@@ -0,0 +1,157 @@
+= ServerSide Key Generation for EE
+Author <cfu@redhat.com>
+:toc:
+:imagesdir: images
+
+== Overview
+
+Many newer versions of browsers, including Firefox v69 and up, as well as Chrome, have removed the functionality to generate PKI keys and the support for CRMF for key archival. While CLIs such as CRMFPopClient (see "CRMFPopClient --help") or pki (see "pki client-cert-request --help") could be used as a workaround on the Fedora or RHEL platforms, clients on non-RedHat platforms have been largely left out. 
+
+Server-Side Keygen enrollment has been around for a long time since the introduction of Token Key Management System (TMS), where keys could be generated on KRA instead of locally on smart cards. In this release, we adopt a similar mechanism to resolve the browser keygen deficiency issue.   Keys are generated on the server (KRA to be specific) and then transferred securely back to the client in PKCS#12.
+
+NOTE: it is highly recommended that the Server-Side Keygen mechanism only being employed for encryption certificates.
+
+== Functionality Highlights
+
+* Certificate request keys are generated on KRA (Note: KRA must be installed to work with the CA)
+* The profile default plugin, serverKeygenUserKeyDefaultImpl, provides selection to enable or disable key archival (i.e. the “enableArchival” parameter)
+* Support for both RSA and EC keys
+* Support for both manual (agent) approval and automatic approval (e.g. directory password-based)
+
+== Installation configuration
+
+A KRA instance is required in addition to the CA for setting up Server-Side Keygen.
+
+After installing CA and KRA instances, in the case of stand-alone Tomcat web server instances, you’d need to add the KRA’s transport cert to the CA’s nssdb.
+
+First, shut down CA
+where SERVER_SIDE_KEYGEN_ENROLL_KEY_RETRIEVAL_REQUEST_PROCESSED is not yet added
+[literal]
+Systemctl stop pki-tomcatd@instance_name.service
+
+Export KRA’s transport certificate into a file
+
+[literal]
+certutil -d /var/lib/pki/instance_name/kra/alias -L -n transport-cert-nickname -a > kra.cert
```

Comment:
The line before that says "grep "ca.connector.KRA.transportCertNickname" /var/lib/pki/instance_name/ca/conf/CS.cfg".
Though I think I should use <kra_instance_name> as you suggested for other areas.

Reply:
oops, I meant <ca_instance_name>

Reply:
Note that there are two `certutil` commands. The first one exports the transport cert from KRA and the second one imports it into CA. There's only one `grep` command between the two `certutil` commands that grabs the transport cert nickname from CA's CS.cfg. Are the transport cert nicknames in CA and KRA guaranteed to be the same? If not, we might need a separate `grep` command for KRA.

Also, logically the admin will have to run `grep` first before `certutil`, so the steps in the doc should also be written in the same order.

Reply:
The first certutil is to export the kra transport cert from the token.  I guess I could add a certutil -d /var/lib/pki/<kra_instance_name>/kra/alias -L to display all certs in the token.

Reply:
Is it possible to `grep` the transport cert nickname from KRA's CS.cfg? That would be more accurate than listing all certs in KRA using `certutil` and ask the admin to pick one. If the KRA uses an HSM there might be many certs with similar nicknames.

Reply:
sure.

Reply:
looks like there's a kra transport cert in the CS.cfg. I could find it with
 grep "kra.transport.cert=" CS.cfg > kra.cert
but is there a way to get rid of the  "kra.transport.cert=" from the output so kra.cert will only contain the b64 of the cert itself?

Reply:
In PKI 10.9 (or maybe earlier) you can get the nicknames with this command:
```
$ pki-server cert-show -i <ca_instance_name> ca_signing
  Cert ID: signing
  Nickname: edewata/ca_signing
  ...
$ pki-server cert-show -i <kra_instance_name> kra_transport
  Cert ID: transport
  Nickname: edewata/kra_transport
  ...
```
I don't think the command is available in PKI 10.5 though, so you probably need to use something like this:
```
$ grep "kra.transport.cert=" CS.cfg | sed 's/kra.transport.cert=//'
```
But are you looking for the nickname? There are `kra.cert.transport.nickname`, `kra.transport.nickname`, and `kra.transportUnit.nickName` params in the CS.cfg.

Reply:
The previous method was to find the kra transport nickname, then export the cert from the nssdb;  Now that since we found the transport cert itself in the CS.cfg, the grep/sed command you gave is perfect as a single step solution!  Thanks.